src/Integration.UnitTests/Binding/BindingProcessImplTests.cs

```diff
@@ -256,33 +205,77 @@ public async Task DownloadQualityProfile_SavesConfiguration()
+            serviceProvider.RegisterService(typeof(IConfigurationPersister), configPersister);
 
-            // CSharp is a supported language, but only the VB.NET QP is available
-            var language = Language.CSharp;
-            host.SupportedPluginLanguages.Add(language);
+            var testSubject = CreateTestSubject(
+                configProvider: configProvider.Object,
+                languagesToBind: languagesToBind);
 
-            this.ConfigureQualityProfile(Language.VBNET, "");
+            var notifications = new ConfigurableProgressStepExecutionEvents();
+            var progressAdapter = new FixedStepsProgressAdapter(notifications);
 
             // Act
             var result = await testSubject.DownloadQualityProfileAsync(progressAdapter, CancellationToken.None);
 
             // Assert
-            result.Should().BeFalse();
-            testSubject.InternalState.BindingConfigs.Should().NotContainKey(Language.VBNET, "Not expecting any rules for this language");
-            testSubject.InternalState.BindingConfigs.Should().NotContainKey(language, "Not expecting any rules");
+            result.Should().BeTrue();
+            testSubject.InternalState.BindingConfigs.Should().NotContainKey(Language.Cpp);
+            testSubject.InternalState.BindingConfigs.Should().NotContainKey(Language.Secrets);
+            testSubject.InternalState.BindingConfigs.Should().NotContainKey(Language.Css);
 
-            notifications.AssertProgressMessages(Strings.DownloadingQualityProfileProgressMessage);
+            testSubject.InternalState.BindingConfigs.Should().ContainKey(Language.CSharp);
+            testSubject.InternalState.BindingConfigs.Should().ContainKey(Language.VBNET);
 
-            logger.AssertOutputStrings(1);
-            var expectedOutput = string.Format(Strings.SubTextPaddingFormat,
-                string.Format(Strings.CannotDownloadQualityProfileForLanguage, language.Name));
-            logger.AssertOutputStrings(expectedOutput);
+            // Progess notifications - percentage complete and messages
+            notifications.AssertProgress(0.25, 0.5, 0.75, 1.0);
+            CheckProgressMessages(languagesToBind);
+
+            // Check output messages
+            var missingPluginMessageCFamily = string.Format(Strings.SubTextPaddingFormat,
+                string.Format(Strings.CannotDownloadQualityProfileForLanguage, Language.Cpp.Name));
+            var missingPluginMessageSecrets = string.Format(Strings.SubTextPaddingFormat,
+                string.Format(Strings.CannotDownloadQualityProfileForLanguage, Language.Secrets.Name));
+
+            logger.AssertOutputStringExists(missingPluginMessageCFamily);
+            logger.AssertOutputStringExists(missingPluginMessageSecrets);
+            logger.AssertPartialOutputStringDoesNotExist(Language.Css.Name);
+
+            void CheckProgressMessages(params Language[] languages)
```

Comment:
Checking the progress messages is done in two nested methods (this one and the one in line 267).
I'm not sure the second nested method helps readability or not, since it's only one line long.

Reply:
hmm I think it is okay this way, it did take me a tiny bit too understand but not a significant amount